### ISSUE TITLE
Vacuum command should support "full" mode too!

### ISSUE THREAD

**Issue by bnewman-tech** [delta-io/delta-rs#2349]
# Environment

**Delta-rs version**: 0.15.3

**Binding**: Python

**Environment**: 
Local Machine - MacBook M1
We are using Azure Blob Storage.

***
# Bug

**What happened**:
We are seeing an issue where the vacuum command is not removing all the old files. 

Here you can see we ran the vacuum command in Databricks and there are 137 files to remove. 
![image](https://github.com/delta-io/delta-rs/assets/60909819/ab257399-16d7-4a49-b270-62feb17a299b)
 
Using Delta-RS there was only 4 to remove. 
```
       dt = await get_delta_table(storage_account_path)
        result = dt.vacuum(retention_hours=720, dry_run=True)
        print(f"Number of files to vacuum: {len(result)}")
        for file in result:
            print(file)
```
Number of files to vacuum: 4
345-02f7d21b-7097-4dda-ae68-8e1c11703e53-0.parquet
346-9a154da0-b3f1-4904-be42-4ef4fdf610bc-0.parquet
347-37b41632-8304-4764-ae4d-82effe20371e-0.parquet
350-b0a9041a-13a5-4bfd-91f6-bf629d1191d9-0.parquet

**What you expected to happen**:

We except it to remove all files as stated in the return of the function. 
"the list of files no longer referenced by the Delta Table and are older than the retention threshold."

Would there be a specific reason for this? 
I am available for further clarification or to provide additional details if needed.

**Comment by ion-elgreco** [comment]
Can you share the table configuration? 

Perhaps your table retention duration is higher than 720 hours? If so then you need to set: `enforce_retention_duration=True`

**Comment by bnewman-tech** [comment]
Thank you for the quick reply!

I've checked the table configuration, and here are the details relevant to the retention duration:

**delta.logRetentionDuration**: 'interval 30 days'

Here's the output of the Delta-RS metadata:
```
Metadata(id: 26af3ffa-dab4-4d78-a611-a9af23a75546, name: None, description: None, partition_columns: [], created_time: 1689193309750, configuration: {'delta.autoOptimize.autoCompact': 'true', 'delta.logRetentionDuration': 'interval 30 days', 'delta.autoOptimize.optimizeWrite': 'true'})
```

After your suggestion, I included enforce_retention_duration=True in the vacuum command like this:
```
result = dt.vacuum(retention_hours=720, dry_run=True, enforce_retention_duration=True)
```
It also looks like enforce_retention_duration is set to True by default. 

Unfortunately, the result was the same, with only 4 files identified for removal.

Thank you for your assistance!

**Comment by ion-elgreco** [comment]
@bnewman-tech what do you see if you do this: 
`dt.vacuum(retention_hours=0, dry_run=True, enforce_retention_duration=False)`

**Comment by bnewman-tech** [comment]
I get the following. 

```
Number of files to vacuum: 34
345-02f7d21b-7097-4dda-ae68-8e1c11703e53-0.parquet
346-9a154da0-b3f1-4904-be42-4ef4fdf610bc-0.parquet
347-37b41632-8304-4764-ae4d-82effe20371e-0.parquet
350-b0a9041a-13a5-4bfd-91f6-bf629d1191d9-0.parquet
351-c3608b3d-9d01-4cc6-815f-59b91fabb852-0.parquet
352-6807e211-ee97-447a-873f-2a037b5d45da-0.parquet
353-145f57e6-7fce-43b9-91f7-e5b982432b3a-0.parquet
354-5c29bf90-d95c-4060-abc9-875b9d081953-0.parquet
355-c35779c4-e702-41e6-8cc1-94ef34f4c6d5-0.parquet
356-f58cd1f8-9263-4886-a564-e05329d803ad-0.parquet
359-4062474d-1e18-4d63-8100-7c3e1bca7165-0.parquet
360-28a94bbe-c38e-425b-9025-4426238746e5-0.parquet
361-c885f8af-3b3a-4dd8-93c1-ef5155821c9f-0.parquet
362-45e5469c-5616-4222-b1c8-d0b7f94c9cb8-0.parquet
363-8ac9fdaa-3c59-4c50-ba35-dc7adda933db-0.parquet
364-8949ba16-ed70-4c9f-a810-db2be73e0910-0.parquet
365-f3656dd8-6988-41a9-ba6e-c9ad4c5adca9-0.parquet
368-d6393649-a0ff-421c-9cd8-b1bf4e5650b1-0.parquet
369-5c391218-b16b-4349-af9b-ecb1c859331d-0.parquet
370-3904e9a8-5654-4611-8a65-103de380d0bc-0.parquet
371-6ccdb486-81f2-42db-9a10-5e7ba219d70f-0.parquet
372-168c009b-8ce0-40e9-a9ba-c4981e296bc2-0.parquet
373-4bfe8162-a934-419a-8dea-152c26c49e88-0.parquet
374-0c571bcf-a593-4c78-a7ff-2d9ed4f85a87-0.parquet
377-e37c8103-5ca0-4201-8b17-1ce707d472d9-0.parquet
378-d8cb9b24-d5b9-4b54-a686-cb4a20e139c2-0.parquet
379-7d85ac2f-0194-473c-ad47-1cf927ceb3da-0.parquet
380-bb71b170-2107-48d9-a790-4c2a86128ed7-0.parquet
381-b21c737f-df10-48d6-b79c-616f2694aa5b-0.parquet
382-ce9e6d1b-af8d-475a-b4f4-a022b28dc305-0.parquet
383-7c4f8fd6-52c2-4bd4-ac6f-123fce19019f-0.parquet
386-741c8b47-617e-4524-a514-dd800a5ebab9-0.parquet
387-8684a8ce-c314-4724-a46f-861c683844ae-0.parquet
388-97c91469-7d55-4afa-bb3a-fc43413de3fe-0.parquet
```

**Comment by ion-elgreco** [comment]
Can you check whether the files that are mentioned by spark-delta and not mentioned by delta-rs, whether they are mentioned in any of the add/remove actions of the log files?

**Comment by Tommel71** [comment]
I have ran into a similar issue. The files are not referenced by the table anymore but are not vacuumed when I use 

```
table.vacuum(
            retention_hours=0, enforce_retention_duration=False, dry_run=False
        )
```

It vacuums files that were unreferenced after an optimize.compact run. 

But it doesnt vacuum files that were removed by using:
```
dl.write_deltalake(
                        table_path,
                        table,
                        partition_by=partition_by,
                        mode=delta_write_mode,
                        schema_mode="overwrite",
                        partition_filters=partition_filters,
                        storage_options=storage_options,
                        **options,
                    )
```
The files deleted by this overwrite were mentioned in the removed action, never in the add action of the log files. The above function is supposed to delete the entire partition when using overwrite (as i do) and it also does that correctly when I try it locally, but in our server environment it doesnt work.

locally i only ever tested it on small amounts of data and I tested the function immediately after the ingest. On the server some time has passed between ignestion and vacuum and its more data.

**Comment by ion-elgreco** [comment]
I will give this a look today or tomorrow, it's probably missing some checks

**Comment by Tommel71** [comment]
I can send you a table with the issue (~200MB) if it helps

**Comment by Tommel71** [comment]
> > I can send you a table with the issue (~200MB) if it helps
> 
> If you can create a small reproducible example with pandas or Polars, that would help!

I have tried and couldn't replicate it. I also tried duplicating an item in a partition and renaming it and it is still there after vacuuming.

**Comment by Tommel71** [comment]
One more note: I am not entirely sure how the logs work but the files were marked as "remove" in several of the parquet files but at one point they were not included anymore. Does "unreferenced" mean they are part of the "remove" logs, or does it mean just not referenced by the logs at all?

**Comment by Tommel71** [comment]
Maybe a flag --remove-unreferenced would be good to delete all files from the folders that are not present in the logs. Unfortunately I am still not able to reproduce this behavior.

**Comment by L5RK** [comment]
Any progress or news on this? We are getting a similar issue at my workplace. Unfortunately due to compliance issues I am not able to share details, however any confirmation whether this is a genuine bug would be helpful. I'm not too familiar with rust, but it seems that the vacuum command tracks the jsons in the logs file. In our application, the trail of file references in the jsons seems to not be cut at the point at which the vacuum stops identifying files. Our log retention policy is significantly greater than our vacuum frequency in order to try to resolve the issue, but no luck. 

As Tommel71 suggested, a remove unreferenced flag would be useful and perhaps more resilient in certain contexts. This seems consistent with the databricks implementation for deltalake, which does offer this functionality via the [full vs lite argument](https://docs.databricks.com/aws/en/sql/language-manual/delta-vacuum)

**Comment by ion-elgreco** [comment]
@L5RK no progress, but we are open to PRs

**Comment by ldacey** [comment]
What is the best way to do this manually? List all parquet files for each partitions and compare versus the dt.files() list and then remove the difference? I have some datasets which were using the overwrite partition method so there are hundreds of files per partition (dt.files() only shows 1-2 files per partition after compacting the table).

**Comment by rtyler** [comment]
@L5RK you are correct in referencing the [lite versus full behavior](https://docs.delta.io/latest/delta-utility.html#remove-files-no-longer-referenced-by-a-delta-table) functionality of Delta/Spark here. The delta-rs implementation as far as I currently understand it is the _lite_ behavior rather than _full_.

I have modified this to an enhancement request and will tackle implementing the _full_ behavior here 🤔

**Comment by yuanhh** [comment]
Hi @rtyler ,

I'm also checking this behavior that from some reason an very old version is left on the storage, but the afterward vacuum doesn't count the file in deletion when running vacuum. Is the current implementation read all transaction log to get _all_tombstones_ files?
per my understanding, the lite mode only read the recent or latest commit in transaction log to determine the removed files, so it's probably could never clean up a removed file that old, is it an expect behavior?
